staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go

```diff
@@ -260,22 +286,45 @@ func (t *envelopeTransformer) doEncode(request *kmstypes.EncryptedObject) ([]byt
 	if err := ValidateEncryptedObject(request); err != nil {
 		return nil, err
 	}
-	return proto.Marshal(request)
+
+	encodingBuffer := t.encryptedObjectDecoder.bufferPool.Get().(*proto.Buffer)
+	defer t.encryptedObjectDecoder.bufferPool.Put(encodingBuffer)
+
+	encodingBuffer.Reset()
+	if err := encodingBuffer.Marshal(request); err != nil {
+		return nil, err
+	}
+
+	return encodingBuffer.Bytes(), nil
 }
 
 // doDecode decodes the byte array to an EncryptedObject.
-func (t *envelopeTransformer) doDecode(originalData []byte) (*kmstypes.EncryptedObject, error) {
-	o := &kmstypes.EncryptedObject{}
-	if err := proto.Unmarshal(originalData, o); err != nil {
+func (t *envelopeTransformer) doDecode(originalData []byte, o *kmstypes.EncryptedObject) (*kmstypes.EncryptedObject, error) {
+	buffer := proto.NewBuffer(originalData)
+
+	resetKMSv2EncryptedObject(o)
+	if err := buffer.Unmarshal(o); err != nil {
```

Comment:
No need for the `buffer` here:

```suggestion
	resetKMSv2EncryptedObject(o)
	if err := proto.Unmarshal(originalData, o); err != nil {
```